server/app/services/program/ActiveAndDraftPrograms.java

```diff
@@ -70,30 +80,110 @@ private ImmutableMap<String, ProgramDefinition> mapNameToProgram(
     // an additional database lookup in order to sync the set of questions associated with the
     // program.
+
+    // Active, non-disabled programs.
     ImmutableMap<String, ProgramDefinition> activeNameToProgram =
+        mapNameToProgramWithFilter(repository, service, active, Optional.of(DisplayMode.DISABLED));
+
+    // All active programs (including disabled).
+    ImmutableMap<String, ProgramDefinition> activeNameToProgramAll =
         mapNameToProgram(repository, service, active);
 
+    // Draft, non-disabled programs.
     ImmutableMap<String, ProgramDefinition> draftNameToProgram =
+        mapNameToProgramWithFilter(repository, service, draft, Optional.of(DisplayMode.DISABLED));
+
+    // All draft programs (including disabled).
+    ImmutableMap<String, ProgramDefinition> draftNameToProgramAll =
         mapNameToProgram(repository, service, draft);
 
-    this.activePrograms = activeNameToProgram.values().asList();
-    this.draftPrograms = draftNameToProgram.values().asList();
-    this.versionedByName =
-        Sets.union(activeNameToProgram.keySet(), draftNameToProgram.keySet()).stream()
-            .collect(
-                ImmutableMap.toImmutableMap(
-                    Function.identity(),
-                    programName -> {
-                      return Pair.create(
-                          Optional.ofNullable(activeNameToProgram.get(programName)),
-                          Optional.ofNullable(draftNameToProgram.get(programName)));
-                    }));
+    if (types.size() < 3) {
+      if (types.contains(ActiveAndDraftProgramsType.DISABLED)
+          && types.contains(ActiveAndDraftProgramsType.IN_USE)) {
+        this.activePrograms = activeNameToProgramAll.values().asList();
+        this.draftPrograms = draftNameToProgramAll.values().asList();
+        this.versionedByName =
+            createVersionedByNameMap(activeNameToProgramAll, draftNameToProgramAll);
+      } else if (types.size() == 1) {
+        ActiveAndDraftProgramsType type = types.get(0);
+        this.activePrograms = activeNameToProgram.values().asList();
+        this.draftPrograms = draftNameToProgram.values().asList();
+        if (type.equals(ActiveAndDraftProgramsType.DISABLED)) {
+          // Disabled active programs.
+          ImmutableMap<String, ProgramDefinition> disabledActiveNameToProgram =
+              filterMapNameToProgram(activeNameToProgramAll, activeNameToProgram);
+          // Disabled draft programs.
+          ImmutableMap<String, ProgramDefinition> disabledDraftNameToProgram =
+              filterMapNameToProgram(draftNameToProgramAll, draftNameToProgram);
+          this.versionedByName =
+              createVersionedByNameMap(disabledActiveNameToProgram, disabledDraftNameToProgram);
+        } else if (type.equals(ActiveAndDraftProgramsType.IN_USE)) {
+          this.versionedByName = createVersionedByNameMap(activeNameToProgram, draftNameToProgram);
+        } else {
```

Comment:
**suggestion:** We should be able to remove these illegal argument exceptions because it won't be possible to add an invalid type into the EnumSet